android/app/src/test/kotlin/coop/polypoly/polypod/updatenotification/UpdateNotificationTest.kt

```diff
@@ -0,0 +1,91 @@
+package coop.polypoly.polypod.updatenotification
+
+import androidx.test.core.app.ApplicationProvider
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.google.common.truth.Truth.assertThat
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.annotation.Config
+import org.robolectric.annotation.LooperMode
+
+@LooperMode(LooperMode.Mode.PAUSED)
+@RunWith(AndroidJUnit4::class)
+@Config(sdk = [Config.OLDEST_SDK])
+class UpdateNotificationTest {
+    @Test
+    fun `notification with id = 0 considered seen`() {
+        val notification = loadNotification(0)
+        assertThat(notification.showInApp).isFalse()
+        assertThat(notification.showPush).isFalse()
+    }
+
+    @Test
+    fun `first ever notification considered not seen`() {
+        val notification = loadNotification(1)
+        assertThat(notification.showInApp).isTrue()
+        assertThat(notification.showPush).isTrue()
+    }
+
+    @Test
+    fun `previously shown notification considered seen`() {
+        val notification = loadNotification(1)
+        notification.onInAppSeen()
+
+        val secondNotification = loadNotification(1)
+        assertThat(secondNotification.showInApp).isFalse()
+        assertThat(secondNotification.showPush).isFalse()
+    }
+
+    @Test
+    fun `additional notification with different id considered not seen`() {
+        val notification = loadNotification(1)
+        notification.onInAppSeen()
+
+        val secondNotification = loadNotification(2)
+        assertThat(secondNotification.showInApp).isTrue()
+        assertThat(secondNotification.showPush).isTrue()
+    }
+
+    @Test
+    fun `additional notification with lower id considered not seen`() {
+        // This is more accidental than desired behaviour - it may make
+        // sense to change the logic so that only notifications with a _higher_
+        // ID than the previous one are shown. But so far we had not reason to.
```

Comment:
I couldn't really think of a problem with this, but if anyone can, happy to change the logic.

The main difference I can think about is that with the current implementation, the same notification _could_ in theory show up twice. Bear with me:

1. We run a notification with ID `1`, that says `First hello` to the user. It shows up, they close it, it's marked as seen.
2. We run a notification with ID `2`, that says `Second hello` to the user. It shows up, they close it, marked as seen.
3. We the first notification _again_ - it will show up again.

A similar case could occur when in step 2, instead of showing a _new_ notification, we disable notifications by setting the ID to `0`. If we set it back again, the new notification ID would be different from `0`, which means it shows. Even if the user already saw it.

All of these use cases imply a mistake on our end, but still, if you think it makes sense I could change the logic so that notifications only get shown if their ID is _higher_ than the previously seen notification. That way there is no way back, only forward. This could lead to problems if we ever accidentally give a notification an ID of `MAX_INT` -> We'd never be able to show another one again :D

Guess there's no way to fully protect ourselves against our own (potential) stupidity.

Reply:
You could use a companion object as a counter, and ditch the 0 as a marker of notification seen. But other alternatives for notification ID are possible, simply use a hash from the title and text.

Reply:
The problem is that you allow the client to set arbitrary IDs, or at least arbitrary mock IDs. So you give them ammo to shoot themselves in the foot.

Reply:
Yeah, the whole _mock_ thing bothers me, but I can't quite think of another solution to make this count testable... But yeah, `MockData` is a footgun for sure.

Note that `0` doesn't mean _this notification was seen_, it's rather a special ID that means notifications are disabled completely. We've used that quite a lot, as the polyPod (up until now) shipped with a notification with ID `0`, since we never wanted to show one :D

I pondered some more whether notifications should only be shown if their ID is _higher_ than the previous one, and I think I'd feel safer with that implementation! Shouldn't make a real difference in practice, but one less thing to worry about I guess, I'll change it.